rama-net/src/client/pool.rs

```diff
@@ -0,0 +1,577 @@
+    H: ReqToConnHasher<Request>,
+{
+    type Response = EstablishedClientConnection<LeasedConnection<S::Connection>, State, Request>;
+    type Error = BoxError;
+
+    async fn serve(
+        &self,
+        ctx: Context<State>,
+        req: Request,
+    ) -> Result<Self::Response, Self::Error> {
+        let (req, hash) = self.hasher.hash(req);
+
+        let pool_result = if let Some(duration) = self.wait_for_pool_timeout {
+            timeout(duration, self.pool.get_connection_or_create_cb(&hash))
+                .await
+                .map_err(|err| OpaqueError::from_std(err))?
+        } else {
+            self.pool.get_connection_or_create_cb(&hash).await
+        }?;
+
+        let (ctx, req, leased_conn) = match pool_result {
+            GetConnectionOrCreate::LeasedConnection(leased_conn) => (ctx, req, leased_conn),
+            GetConnectionOrCreate::AddConnection(cb) => {
+                let EstablishedClientConnection { ctx, req, conn } =
+                    self.inner.connect(ctx, req).await.map_err(Into::into)?;
+                let leased_conn = cb(conn);
+                (ctx, req, leased_conn)
+            }
+        };
+
+        Ok(EstablishedClientConnection {
+            ctx,
+            req,
+            conn: leased_conn,
+        })
+    }
+}
+
+#[cfg(test)]
+
+mod tests {
+    use super::*;
+    use crate::client::EstablishedClientConnection;
+    use rama_core::{Context, Service};
+    use std::{
+        convert::Infallible,
+        ops::Add,
+        sync::{Arc, Mutex},
+    };
+    use tokio_test::assert_err;
+
+    struct TestService {
+        pub created_connection: Arc<Mutex<u32>>,
+    }
+
+    impl Default for TestService {
+        fn default() -> Self {
+            Self {
+                created_connection: Arc::new(Mutex::new(0)),
+            }
+        }
+    }
+
+    impl<State, Request> Service<State, Request> for TestService
+    where
+        State: Clone + Send + Sync + 'static,
+        Request: Send + 'static,
+    {
+        type Response = EstablishedClientConnection<Vec<u32>, State, Request>;
+        type Error = Infallible;
+
+        async fn serve(
+            &self,
+            ctx: Context<State>,
+            req: Request,
+        ) -> Result<Self::Response, Self::Error> {
+            let mut counter = self.created_connection.lock().expect("mutex not poisened");
+            *counter = counter.add(1);
+
+            let conn = vec![];
+            Ok(EstablishedClientConnection { ctx, req, conn })
+        }
+    }
+
+    struct StringRequestHasher;
+
+    impl ReqToConnHasher<String> for StringRequestHasher {
+        fn hash(&self, req: String) -> (String, String) {
+            let count = req.chars().count().to_string();
+            return (req, count);
+        }
+    }
+
+    #[tokio::test]
+    async fn test_should_reuse_connections() {
+        let pool = Pool::new(NonZeroU16::new(1).unwrap(), NonZeroU16::new(1).unwrap()).unwrap();
```

Comment:
might want to provide somewhere in rama-utils a macro or something else to make this a bit more user-friendly to use, e.g.:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=cc2e04792fef1d5599a216c58a5a6590

Only really required for literals as in the case you deserialize it (e.g. for a config) than it is already checked when deserializing.